### DISCUSSION TITLE
Call external tools to translate the selection

### DISCUSSION THREAD

**Discussion by QiWei** [sumatrapdfreader/sumatrapdf#3214] [Ideas]
The translator does not need to open a browser and sends a direct http request so that the selected content can be translated by calling an external tool through the API.

**Comment by GitHubRulesOK** [comment]
there are many examples of translate via local external application without use browser but browser is the most universal so added to recent versions so earlier methods required vbs or ahk to use clipboard as the intermediary to web or internal applications  https://forum.sumatrapdfreader.org/t/passing-selection-to-an-app-e-g-web-search-3-4-pre-release/2700

**Comment by QiWei** [comment]
The old program with VBS analog keys might as well have used AutoHotkey instead.
The disadvantage of the new solution is that the web page is opened once per request, which is not friendly to users who are slow to access the translation site.
In fact, it is possible to combine the two solutions, that is, you can send the selected text as a parameter to an external tool.
The configuration would look like the following:
```
SelectionHandlers [
	[
		CMD = @curl -sS "http://127.0.0.1:11111/lingoes?cmd=main_search&text=$${selection}"
		Name = Translate with Lingoes
	]
]
```
If the external tool is preceded by the '@' symbol, it means that it runs in a hidden way.

**Reply by GitHubRulesOK** [reply]
Hmm I don't run a local server to test APIs but potentially that should be URL based.
```
SelectionHandlers [
	[
		URL = http://127.0.0.1:11111/lingoes?cmd=main_search&text=$${selection}
		Name = Translate with Lingoes
	]
]
```
